src/control/server/server.go

```diff
@@ -38,437 +32,324 @@ import (
-		req.SetHostList(cfg.AccessPoints)
-		req.SetSystem(cfg.SystemName)
-		req.ControlAddr = controlAddr
-		return control.SystemJoin(ctx, rpcClient, req)
+func (srv *server) initStorage() error {
+	runningUser, err := user.Current()
+	if err != nil {
+		return errors.Wrap(err, "unable to lookup current user")
 	}
 
-	var allStarted sync.WaitGroup
+	if err := prepBdevStorage(srv, runningUser, iommuDetected(), getHugePageInfo); err != nil {
+		return err
+	}
 
-	for idx, engineCfg := range cfg.Engines {
-		// Provide special handling for the ofi+verbs provider.
-		// Mercury uses the interface name such as ib0, while OFI uses the
-		// device name such as hfi1_0 CaRT and Mercury will now support the
-		// new OFI_DOMAIN environment variable so that we can specify the
-		// correct device for each.
-		if strings.HasPrefix(engineCfg.Fabric.Provider, "ofi+verbs") && !engineCfg.HasEnvVar("OFI_DOMAIN") {
-			deviceAlias, err := netdetect.GetDeviceAlias(netCtx, engineCfg.Fabric.Interface)
-			if err != nil {
-				return errors.Wrapf(err, "failed to resolve alias for %s", engineCfg.Fabric.Interface)
-			}
-			envVar := "OFI_DOMAIN=" + deviceAlias
-			engineCfg.WithEnvVars(envVar)
-		}
+	return srv.ctlSvc.Setup()
+}
 
-		// If the configuration specifies that we should explicitly set
-		// hugepage values per engine instance, do it. Otherwise, let
-		// SPDK/DPDK figure it out.
-		if cfg.SetHugepages {
-			// If we have multiple engine instances with block devices, then
-			// apportion the hugepage memory among the instances.
-			engineCfg.Storage.Bdev.MemSize = hugePages.FreeMB() / len(cfg.Engines)
-			// reserve a little for daos_admin
-			engineCfg.Storage.Bdev.MemSize -= engineCfg.Storage.Bdev.MemSize / 16
-		}
+func (srv *server) createEngine(ei int, ec *engine.Config) (*EngineInstance, error) {
+	// Closure to join an engine instance to a system using control API.
+	joinFn := func(ctxIn context.Context, req *control.SystemJoinReq) (*control.SystemJoinResp, error) {
+		req.SetHostList(srv.cfg.AccessPoints)
+		req.SetSystem(srv.cfg.SystemName)
+		req.ControlAddr = srv.ctlAddr
 
-		// Indicate whether VMD devices have been detected and can be used.
-		engineCfg.Storage.Bdev.VmdDisabled = bdevProvider.IsVMDDisabled()
+		return control.SystemJoin(ctxIn, srv.mgmtSvc.rpcClient, req)
+	}
 
-		bp, err := bdev.NewClassProvider(log, engineCfg.Storage.SCM.MountPoint, &engineCfg.Storage.Bdev)
+	// Indicate whether VMD devices have been detected and can be used.
+	ec.Storage.Bdev.VmdDisabled = srv.bdevProvider.IsVMDDisabled()
+
+	// TODO: ClassProvider should be encapsulated within bdevProvider
+	bcp, err := bdev.NewClassProvider(srv.log, ec.Storage.SCM.MountPoint, &ec.Storage.Bdev)
+	if err != nil {
+		return nil, err
+	}
+
+	engine := NewEngineInstance(srv.log, bcp, srv.scmProvider, joinFn,
+		engine.NewRunner(srv.log, ec)).WithHostFaultDomain(srv.harness.faultDomain)
+
+	if ei == 0 {
+		return engine, nil
+	}
+
+	configureFirstEngine(engine, srv.sysdb, joinFn)
```

Comment:
Should this logic be reversed? Isn't instance 0 the first engine? Maybe I'm missing something.

Reply:
+1 I think this is backwards.